Describe best way to organize screen changes (#529)
smartdevicelink/sdl_app_library_guides: docs/Displaying a User Interface/Main Screen Templates/index.md

```
!@

@![iOS]
When changing screen layouts or template fields, it is highly recommended to encapsulate these updates into a class or method. Doing so is a good way to keep SDL UI changes organized. An example of this can be seen in the [example weather app](https://github.com/SmartDeviceLink-Examples/example_weather_app_ios). See below for another example:

> |~

Reviewer: This code block is too long and has no explanatory comments. I think we need to split this up, but doing so for both objc and swift will be a bit tricky since swift is one file and objc is two.

1. I think we can definitely separate the protocol. We should have a small (markdown) title with a paragraph before it explaining what it is and does.
2. The home screen / button can be another one, again with a markdown title. The objc `.h` and `.m` would be in one tab and the swift in another.
3. Please add stubs with `// TODO` for JavaScript and Java Suite.

```objc
// CustomSDLScreen.h - Protcol
@class SDLManager;

@protocol CustomSDLScreen <NSObject>

@required
- (instancetype)initWithManager:(SDLManager *)sdlManager;
- (void)showScreen;

@end

// HomeSDLScreen.h
#import <Foundation/Foundation.h>
#import "CustomSDLScreen.h"

@class HomeDataModel;
@class SDLManager;
@class SDLScreenManager;

NS_ASSUME_NONNULL_BEGIN

@interface HomeSDLScreen : NSObject<CustomSDLScreen>

@end

NS_ASSUME_NONNULL_END

> // HomeSDLScreen.m

Reviewer: ```suggestion
/************************************************************************/
// HomeSDLScreen.m
```

#import "HomeSDLScreen.h"

#import "CustomSDLScreen.h"
#import "HomeDataModel.h"
#import "SmartDeviceLink.h"

@interface HomeSDLScreen()

@property (strong, nonatomic) SDLManager *sdlManager;

> @property (assign, nonatomic) HomeDataModel *homeDataModel;

Reviewer: ```suggestion
// An example of your data model that will feed data to the SDL screen's UI
@property (strong, nonatomic) HomeDataViewModel *homeDataViewModel;
```

The way you're using this is really more like a view model than a model. Please rename it to reflect that.

@end

@implementation HomeSDLScreen

- (instancetype)initWithManager:(SDLManager *)sdlManager {
    _sdlManager = sdlManager;
    _homeDataModel = HomeDataModel();
}

- (void)showScreen {
    [self.sdlManager.screenManager beginUpdates];
    [self.sdlManager.screenManager changeLayout:[[SDLTemplateConfiguration alloc] initWithTemplate:SDLPredefinedLayoutGraphicWithText] withCompletionHandler:nil];
    self.sdlManager.screenManager.textField1 = self.homeDataModel.text1;
    self.sdlManager.screenManager.textField2 = self.homeDataModel.text2;
    self.sdlManager.screenManager.textField3 = self.homeDataModel.text3;
    self.sdlManager.screenManager.textField4 = self.homeDataModel.text4;

>     self.sdlManager.screenManager.softButtonObjects = @[];

Reviewer: Add buttons that show the other screen.

@property (assign, nonatomic) ButtonDataModel *buttonDataModel;

@end

@implementation ButtonSDLScreen

- (instancetype)initWithManager:(SDLManager *)sdlManager {
    _sdlManager = sdlManager;
    _buttonDataModel = ButtonDataModel();
}

- (void)showScreen {
    [self.sdlManager.screenManager beginUpdates];
    [self.sdlManager.screenManager changeLayout:[[SDLTemplateConfiguration alloc] initWithTemplate:SDLPredefinedLayoutTilesOnly] withCompletionHandler:nil];
    self.sdlManager.screenManager.textField1 = self.buttonDataModel.text1;
    self.sdlManager.screenManager.textField2 = self.buttonDataModel.text2;
    self.sdlManager.screenManager.textField3 = self.buttonDataModel.text3;
    self.sdlManager.screenManager.textField4 = self.buttonDataModel.text4;
    self.sdlManager.screenManager.softButtonObjects = self.buttonDataModel.buttons;
    [self.sdlManager.screenManager endUpdates];
}

@end

// DataSDLScreen.h
#import <Foundation/Foundation.h>
#import "CustomSDLScreen.h"

@class DataModel;
@class SDLManager;
@class SDLScreenManager;

NS_ASSUME_NONNULL_BEGIN

@interface DataSDLScreen : NSObject<CustomSDLScreen>

@end

NS_ASSUME_NONNULL_END

> // DataSDLScreen.m

Reviewer: What's the difference between this and the home view model? I think this could be removed.

### Screen Change Example Interface
All screens will need to have access to the @![iOS]`SDLScreenManager`!@@![android, javaEE, javaSE, javascript]`ScreenManager`!@ object and a function to display the screen. Therefore, it is recommended to create a generic interface for all screens to follow. For the example below, the `CustomSDLScreen` protocol requires an initializer with the parameters `SDLManager` and a `showScreen` method.

@![iOS]
|~
```objc
// CustomSDLScreen.h
@class SDLManager;

@protocol CustomSDLScreen <NSObject>

@required
- (instancetype)initWithManager:(SDLManager *)sdlManager;
- (void)showScreen;

@end
```
```swift
protocol CustomSDLScreen {
    init(sdlManager: SDLManager)
    func showScreen()
}
```
~|
!@

@![android,javaSE,javaEE]
```java
    // TODO
```
!@

@![javascript]
```js
    // TODO
```
!@

### Screen Change Example Implementations
The following example code shows a few implementations of the example screen changing protocol. It is good practice for screen classes is to keep screen data in a view model. Doing so will add a layer of abstraction for exposing public properties and commands to the screen.

> For the example below, the `HomeScreen` class will inherit the `CustomSDLScreen` interface and will have a property of type `HomeDataViewModel`. The screen manager will change its text fields based on the view model's data. In addition, the home screen will also create a navigation button to open the `ButtonSDLScreen` when pressed.

Reviewer: I disagree with this one. I think that the view models are understandable from context and would basically just be stub classes holding the properties that are used on the screen classes. We already have a lot of code here and I'd like to keep us from muddying the waters too much.

@![iOS]
When changing screen layouts and template data (for example, to show a weather hourly data screen vs. a daily weather screen), it is recommended to encapsulate these updates into a class or method. Doing so is a good way to keep SDL UI changes organized. A fully-formed example of this can be seen in the [example weather app](https://github.com/SmartDeviceLink-Examples/example_weather_app_ios). Below is a generic example:
!@

@![android,javaEE,javaSE]
When changing screen layouts and template data (for example, to show a weather hourly data screen vs. a daily weather screen), it is recommended to encapsulate these updates into a class or method. Doing so is a good way to keep SDL UI changes organized. A fully-formed example of this can be seen in the [example weather app](https://github.com/SmartDeviceLink-Examples/example_weather_app_android). Below is a generic example:
!@

@![javascript]
When changing screen layouts and template data (for example, to show a weather hourly data screen vs. a daily weather screen), it is recommended to encapsulate these updates into a class or method. Doing so is a good way to keep SDL UI changes organized. Below is a generic example:
!@

> ### Screen Change Protocol

Reviewer: ```suggestion
## Screen Change Example Code
```

> All screens will need to have access to the ```SDLScreenManager``` object and a function to display the screen. Therefore, it is recommened to create a protocol for all screens to follow. For the example below, the ```CustomSDLScreen``` protocol requires an initializer with the parameters ```SDLManager``` and a ```showScreen``` method.

Reviewer: ```suggestion
This example code creates an interface that can be implemented by various "screens" of your SDL app. This is a recommended design pattern so that you can separate your code to only involve the data models you need. This is just a simple example and your own needs may be different.

### Screen Change Example Interface
All screens will need to have access to the @![iOS]`SDLScreenManager`!@@![android, javaEE, javaSE, javascript]`ScreenManager`!@ object and a function to display the screen. Therefore, it is recommended to create a generic interface for all screens to follow. For the example below, the `CustomSDLScreen` protocol requires an initializer with the parameters `SDLManager` and a `showScreen` method.
```

@![iOS]
|~
```objc
// CustomSDLScreen.h
@class SDLManager;

@protocol CustomSDLScreen <NSObject>

@required
- (instancetype)initWithManager:(SDLManager *)sdlManager;
- (void)showScreen;

@end
```
```swift
protocol CustomSDLScreen {
    init(sdlManager: SDLManager)
    func showScreen()
}
```
~|
!@

@![android,javaSE,javaEE]
```java
    // TODO
```
!@

@![javascript]
```js
    // TODO
```
!@

> ### Screen Classes

Reviewer: ```suggestion
### Screen Change Example Implementations
```

> As previously mentioned most screens should all follow a shared protocol, such as the ```CustomSDLScreen``` protocol above, since screens will likely operate in a similar fashion. Another good practice for screen classes is to keep screen data in a view model. Doing so will add a layer of abstraction for exposing public properties and commands to the screen. 

Reviewer: ```suggestion
The following example code shows a few implementations of the example screen changing protocol. It is good practice for screen classes is to keep screen data in a view model. Doing so will add a layer of abstraction for exposing public properties and commands to the screen.
```

> For the example below, the ```HomeScreen``` class will inherit the ```CustomSDLScreen``` protocol and will have a property of ```HomeDataViewModel```. The screen manager will change its text fields based on the view model's data. In addition, the home screen will also create a navigation button to open the ```SDLButtonScreen``` when pressed.

Reviewer: ```suggestion
For the example below, the `HomeScreen` class will inherit the `CustomSDLScreen` interface and will have a property of type `HomeDataViewModel`. The screen manager will change its text fields based on the view model's data. In addition, the home screen will also create a navigation button to open the `ButtonSDLScreen` when pressed.
```

@![iOS]
|~
```objc
// HomeSDLScreen.h
#import <Foundation/Foundation.h>
#import "CustomSDLScreen.h"

@class HomeDataViewModel;
@class SDLManager;
@class SDLScreenManager;
@class SDLSoftButtonObject;

NS_ASSUME_NONNULL_BEGIN

@interface HomeSDLScreen : NSObject<CustomSDLScreen>

@end

NS_ASSUME_NONNULL_END

/************************************************************************/
// HomeSDLScreen.m
#import "HomeSDLScreen.h"

#import "ButtonSDLScreen.h"
#import "CustomSDLScreen.h"
#import "HomeDataViewModel.h"
#import "SmartDeviceLink.h"

@interface HomeSDLScreen()

> @property (strong, nonatomic) SDLManager *sdlManager;

Reviewer: ```suggestion
@property (weak, nonatomic) SDLManager *sdlManager;
```

I suspect this would be weak and held strongly in a central `ProxyManager`

// A button to navigate to the ButtonSDLScreen
@property (strong, nonatomic) SDLSoftButtonObject *navigationButton;
// An example of your data model that will feed data to the SDL screen's UI
@property (strong, nonatomic) HomeDataViewModel *homeDataViewModel;

@end

@implementation HomeSDLScreen

- (instancetype)initWithManager:(SDLManager *)sdlManager {
    self = [super init];
    if (!self) { return nil; }

    _sdlManager = sdlManager;
    // Set navigation button to an SDLSoftButtonObject that will display the Button SDL Screen
    _navigationButton = [[SDLSoftButtonObject alloc] initWithName:@"ButtonSDLScreen" text:@"Button Screen" artwork:nil handler:^(SDLOnButtonPress * _Nullable buttonPress, SDLOnButtonEvent * _Nullable buttonEvent) {
        if (buttonPress == nil) { return; }

>         ButtonSDLScreen *buttonScreen = [[ButtonSDLScreen alloc] initWithManager:sdlManager];

Reviewer: Create the `ButtonSDLScreen` and hold it in the class, then call `[buttonScreen showScreen]`

        buttonScreen.showScreen();
    }];
    _homeDataViewModel = HomeDataViewModel();

    return self;
}

- (void)showScreen {
    // Batch updates
    [self.sdlManager.screenManager beginUpdates];
    // Change template to Graphics With Text
    [self.sdlManager.screenManager changeLayout:[[SDLTemplateConfiguration alloc] initWithTemplate:SDLPredefinedLayoutGraphicWithText] withCompletionHandler:nil];
    // Assign text fields to view model data
    self.sdlManager.screenManager.textField1 = self.homeDataViewModel.text1;
    self.sdlManager.screenManager.textField2 = self.homeDataViewModel.text2;
    self.sdlManager.screenManager.textField3 = self.homeDataViewModel.text3;
    self.sdlManager.screenManager.textField4 = self.homeDataViewModel.text4;
    // Assign to navigation button property

>     self.sdlManager.screenManager.softButtonObjects = @[self.navigationButton];

Reviewer: I don't think you need to create and hold `self.navigationButton` in the `init`. You can create it in this method (or create a function like `- (SDLSoftButtonObject *)createNavigationButton { }`

            guard let buttonPress = buttonPress else { return }
            
            let buttonSDLScreen = ButtonSDLScreen(sdlManager)
            buttonSDLScreen.showScreen()
        }
    }

    func showScreen() {
        // Batch updates
        self.sdlManager.screenManager.beginUpdates()
        // Change template to Graphics With Text
        self.sdlManager.screenManager.changeLayout(SDLTemplateConfiguration(predefinedLayout: .graphicWithText))
        // Assign text fields to view model data
        self.sdlManager.screenManager.textField1 = self.homeDataViewModel.text1
        self.sdlManager.screenManager.textField2 = self.homeDataViewModel.text2
        self.sdlManager.screenManager.textField3 = self.homeDataViewModel.text3
        self.sdlManager.screenManager.textField4 = self.homeDataViewModel.text4
        // Assign to navigation button property
        self.sdlManager.screenManager.softButtonObjects = [self.navigationButton]
        self.sdlManager.screenManager.endUpdates()
    }
}
```
~|
!@

@![android,javaSE,javaEE]
```java
    // TODO
```
!@

@![javascript]
```js
    // TODO
```
!@

> The ```ButtonSDLScreen``` follows the same patterns as the ```HomeSDLScreen``` but has minor implementation differences. The screen's view model ```ButtonDataViewModel``` contains properties unique to the ```ButtonSDLScreen``` such as text fields and an array of soft button objects. It also changes the template configuration to tiles only.

Reviewer: ```suggestion
The `ButtonSDLScreen` follows the same patterns as the `HomeSDLScreen` but has minor implementation differences. The screen's view model `ButtonDataViewModel` contains properties unique to the `ButtonSDLScreen` such as text fields and an array of soft button objects. It also changes the template configuration to tiles only.
```
